lib/mock-module/mock-module.ts

```diff
@@ -1,72 +1,168 @@
+  } else {
+    result.push(values);
+  }
+  return result;
 };
 
-const isModuleWithProviders = (object: any): object is ModuleWithProviders => typeof object.ngModule !== 'undefined';
-
-export function MockModule(module: Type<NgModule> | ModuleWithProviders): Type<NgModule> {
-  return NgModule(MockIt(module))(class MockedModule {});
-}
+// Checks if an object was decorated by NgModule.
+const isModule = (object: any): object is Type<NgModule> => {
+  const annotations = jitReflector.annotations(object);
+  const ngMetadataNames = annotations.map((annotation) => annotation.__proto__.ngMetadataName);
+  return ngMetadataNames.indexOf('NgModule') !== -1;
+};
 
-const NEVER_MOCK: Array<Type<NgModule>> = [CommonModule];
+// Checks if an object implements ModuleWithProviders.
+const isModuleWithProviders = (object: any): object is ModuleWithProviders => {
+  return typeof object.ngModule !== 'undefined' && isModule(object.ngModule);
+};
 
-function MockIt(module: Type<NgModule> | ModuleWithProviders): IModuleOptions {
+export function MockModule(module: ImportInstance, config: IMockModuleConfig = {}): ImportInstance {
+  // Usually we want to test what we have inside of module we passed here explicitly.
+  // For that we should have config.exposeDeclarations = true.
+  // That means all imports and declarations of the module will be exported.
+  // Other modules from list of imports of the module expose only own exports.
+  // It allow us to see failures in case if we use something what wasn't properly exported in nested modules.
+  if (typeof config.exposeDeclarations !== 'boolean') {
+    config.exposeDeclarations = true;
+  }
   let ngModule: Type<NgModule>;
-  let ngModuleProviders: Provider[] = [];
+  let ngModuleProviders: Provider[] | undefined;
+  let moduleMockPointer: Type<any>;
 
-  // Disassembling module.
   if (isModuleWithProviders(module)) {
     ngModule = module.ngModule;
     if (module.providers) {
-      ngModuleProviders = flatten(module.providers);
+      ngModuleProviders = module.providers;
     }
   } else {
     ngModule = module;
   }
 
   if (NEVER_MOCK.includes(ngModule)) {
-    return ngModule as any;
+    return module as any;
+  }
+
+  // Every module should be mocked only once to avoid errors like:
+  // Failed: Type ...Component is part of the declarations of 2 modules: ...Module and ...Module...
+  const cacheHit = cache.get(ngModule);
+  if (cacheHit) {
+    moduleMockPointer = cacheHit;
+  } else {
+    @NgModule(MockIt(ngModule, config))
+    @MockOf(ngModule)
+    class ModuleMock {
+    }
+
+    moduleMockPointer = ModuleMock;
+    cache.set(ngModule, moduleMockPointer);
+  }
+
+  if (ngModuleProviders) {
+    return {
+      ngModule: moduleMockPointer,
+      providers: flatten(ngModuleProviders).map(mockProvider),
+    };
+  }
+  return moduleMockPointer;
+}
+
+const NEVER_MOCK: Array<Type<NgModule>> = [CommonModule, BrowserModule, BrowserAnimationsModule];
```

Comment:
Skipping `BrowserModule`, `BrowserAnimationsModule` allows us to mock root `AppModule`.